### ISSUE TITLE
Limit the scope of customPropertiesWhitelist

### ISSUE THREAD

**Issue by pento** [WordPress-Coding-Standards/WordPress-Coding-Standards#1623]
## Problem

The `customPropertiesWhitelist` property of the `WordPress.NamingConventions.ValidVariableName` sniff is useful for whitelisting variables that can't be changed, but it risks whitelisting some variables incorrectly, as there's no way to limit the scope of when the whitelist should be applied.

For example, when adding `<element value="preserveWhiteSpace"/>` to the property, it'd be useful to only apply that to `DOMDocument` variables.

This should pass:

```php
$doc                     = new DOMDocument();
$doc->preserveWhiteSpace = false;
```

This shouldn't:

```php
$doc                     = new SomeOtherClass();
$doc->preserveWhiteSpace = false;
```

## Possible Solution

As @jrfnl suggested over on [WP45934](https://core.trac.wordpress.org/ticket/45934#comment:17), we could add a `key` property to the `<element>`, allowing a scope to be applied to the whitelist.

`<element key="DOMDocument" value="preserveWhiteSpace"/>`

This becomes a little trickier with functions that can return an object:

```php
$doc   = new DOMDocument();
$xpath = new DOMXPath( $doc );
$rules = $xpath->query( '/configuration/system.webServer/rewrite/rules/rule[starts-with(@name,\'wordpress\')] | /configuration/system.webServer/rewrite/rules/rule[starts-with(@name,\'WordPress\')]' );
if ( $rules->length > 0 ) {
	$child  = $rules->item( 0 );
	// Can we detect that $child is a DOMNode?
	$parent = $child->parentNode;
}
```

Or even worse with functions that can return an arbitrary object:

```php
$index = $wpdb->get_row( "SHOW INDEXES FROM $wpdb->signups" );
if ( 'domain_path' == $index->Key_name ) {
	return;
}
```

**Comment by jrfnl** [comment]
I totally agree that this property could benefit from some sort of scoping, however, when I mentioned:

> We could use the key to limit the whitelisting for a particular property to a specific file/folder or class. 

I didn't mean `class` as "class the property comes from" as that is impossible to determine reliably with a static analysis tool.
Think: an already instantiated object can be passed in as a parameter to a method from within another file. Without type declarations in the method signature - and let's be fair, those are few and far between in WP -, there is no way for the sniff to know what class that instantiated object represents.

There are a few more realistic options:
* Limiting by file name / folder / file regex pattern
    ```xml
    <element key=".*/wp-includes/subdir/.+\.php" value="propertyName"/>
    ```
* Limiting based on the variable object name. This would require some discipline when naming objects or could result in an even larger whitelist
    ```xml
    <element key="$wpdb" value="propertyName"/>
    <element key="$domElm" value="propertyName"/>
    ```
    ```php
    if ($wpdb->propertyName === false) {} // This would be OK.
    if ($db->propertyName === false) {} // This would still be flagged.
    ```
* Limiting based on the name of the class in which the property is used.
    ```xml
    <element key="WPDB" value="propertyName"/>
    <element key="REST_API" value="propertyName"/>
    ```
    ```php
    class WPDB {
        public $propertyName; // This would be OK.
        function test( $obj ) {
            if ( $obj->propertyName ) {} // This would be OK.
        }
    }
    class NotWPDB {
        public $propertyName; // This would still be flagged.
        function test( $obj ) {
            if ( $obj->propertyName ) {} // This would still be flagged.
        }
    }
    ```
    Note: This is different to limiting by file in that WP still contains numerous files with multiple classes.

Now all of these have their shortcomings and could still whitelist too much/be fiddly to work with, but these IMO would be realistic options.
